torchvision/csrc/cpu/video/Video.cpp

```diff
@@ -0,0 +1,349 @@
+}
+
+c10::Dict<std::string, c10::Dict<std::string, std::vector<double, std::allocator<double>>>> Video::getStreamMetadata() const {
+  return streamsMetadata;
+}
+
+void Video::Seek(double ts) {
+  // initialize the class variables used for seeking and retrurn
+  _getDecoderParams(
+      ts, // video start
+      0, // headerOnly
+      get<0>(current_stream), // stream
+      long(get<1>(
+          current_stream)), // stream_id parsed from info above change to -2
+      false // read all streams
+  );
+
+  // calback and metadata defined in Video.h
+  succeeded = decoder.init(params, std::move(callback), &metadata);
+  LOG(INFO) << "Decoder init at seek " << succeeded << "\n" ;
+}
+
+std::tuple<torch::Tensor, double> Video::Next() {
+
+  // if failing to decode simply return a null tensor (note, should we
+  // raise an exeption?)
+  double frame_pts_s;
+  torch::Tensor outFrame = torch::zeros({0}, torch::kByte);
+
+  // decode single frame
+  DecoderOutputMessage out;
+  int64_t res = decoder.decode(&out, decoderTimeoutMs);
+  // if successfull
+  if (res == 0) {
+    frame_pts_s = double(double(out.header.pts) * 1e-6);
+
+    auto header = out.header;
+    const auto& format = header.format;
+
+    // initialize the output variables based on type
+
+    if (format.type == TYPE_VIDEO) {
+      // note: this can potentially be optimized
+      // by having the global tensor that we fill at decode time
+      // (would avoid allocations)
+      int outHeight = format.format.video.height;
+      int outWidth = format.format.video.width;
+      int numChannels = 3;
+      outFrame = torch::zeros({outHeight, outWidth, numChannels}, torch::kByte);
+      auto numberWrittenBytes = fillVideoTensor(out, outFrame);
+      outFrame = outFrame.permute({2, 0, 1});
+
+    } else if (format.type == TYPE_AUDIO) {
+      int outAudioChannels = format.format.audio.channels;
+      int bytesPerSample = av_get_bytes_per_sample(
+          static_cast<AVSampleFormat>(format.format.audio.format));
+      int frameSizeTotal = out.payload->length();
+
+      CHECK_EQ(frameSizeTotal % (outAudioChannels * bytesPerSample), 0);
+      int numAudioSamples =
+          frameSizeTotal / (outAudioChannels * bytesPerSample);
+
+      outFrame =
+          torch::zeros({numAudioSamples, outAudioChannels}, torch::kFloat);
+
+      auto numberWrittenBytes = fillAudioTensor(out, outFrame);
+    }
+    // currently not supporting other formats (will do soon)
+
+    out.payload.reset();
+  } else {
+    LOG(ERROR) << "Decoder failed ( or ran into last iteration)";
+  }
+
+  std::tuple<torch::Tensor, double> result = {outFrame, frame_pts_s};
+  return result;
+}
+
+// Video::~Video() {
+// destructor to be defined thoroughly later
+//   delete params; // does not have destructor
+//   delete metadata; // struct does not have destructor
+//   delete decoder; // should be fine
+//   delete streamFPS; // should be fine
+//   delete streamDuration; // should be fine
+// }
```

Comment:
This might not be needed if there is no `malloc`s in the downstream code

Reply:
Technically, there are calls to `av_malloc` in the decoder init, which is why I think it _might_ make sense to have it, but I haven't had issues so far.